otter/convergence/service.py

```diff
@@ -310,11 +333,24 @@ def converge_one_group(log, group_locks, tenant_id, group_id, version,
     """
     log = log.bind(tenant_id=tenant_id, group_id=group_id)
     eff = execute_convergence(tenant_id, group_id, log)
-    return group_locks(group_id, eff).on(
-        success=lambda r: delete_divergent_flag(log, tenant_id,
-                                                group_id, version),
-        error=partial(_cleanup_convergence_error, log, tenant_id,
-                      group_id, version))
+    try:
+        result = yield group_locks(group_id, eff)
+    except ConcurrentError:
+        # We don't need to spam the logs about this, it's to be expected
+        return
+    except NoSuchScalingGroupError:
+        log.err(None, 'converge-fatal-error')
+        yield delete_divergent_flag(log, tenant_id, group_id, version)
+        return
+    except Exception:
+        # We specifically don't clean up the dirty flag in the case of
+        # unexpected errors, so convergence will be retried.
+        log.err(None, 'converge-non-fatal-error')
+    else:
+        if result in (StepResult.FAILURE, StepResult.SUCCESS):
+            yield delete_divergent_flag(log, tenant_id, group_id, version)
+        # TODO: if result is FAILURE, put the group into ERROR state.
+        # https://github.com/rackerlabs/otter/issues/885
```

Comment:
(Do you have any suggestions? Previously I thought I wanted `TODO: ... (#123)`, figured that was hard to grep for, then maybe `TODO #123: ...`, but I really like that this is a URL and therefore potentially clickable, also that it allows linking to other issues from e.g. other projects.)